core/server/worker/src/main/java/alluxio/worker/block/DefaultBlockWorker.java

```diff
@@ -508,7 +508,14 @@ public BlockReader createUfsBlockReader(long sessionId, long blockId, long offse
       boolean positionShort, Protocol.OpenUfsBlockOptions options)
       throws BlockDoesNotExistException, IOException {
     try {
-      openUfsBlock(sessionId, blockId, options);
+      if (!options.hasUfsPath() && options.hasBlockInUfsTier() && options.getBlockInUfsTier()) {
+        // This is a fallback UFS block read. Reset the UFS block path according to the UfsBlock
+        // flag.
+        UfsManager.UfsClient ufsClient = mUfsManager.get(options.getMountId());
+        options = options.toBuilder()
+            .setUfsPath(alluxio.worker.BlockUtils.getUfsBlockPath(ufsClient, blockId)).build();
+      }
+      mUnderFileSystemBlockStore.acquireAccess(sessionId, blockId, options);
```

Comment:
We can move this to `UnderFileSystemBlockStore.createBlockReader` and make `acquireAccess` private.